src/crafting.cpp

```diff
@@ -940,86 +940,133 @@ void Character::craft_proficiency_gain( const item &craft, const time_duration &
+        // can the NPC actually craft this recipe?
+        bool has_all_skills = np->get_skill_level( making.skill_used ) >= making.difficulty;
+        if( has_all_skills ) {
+            for( const std::pair<const skill_id, int> &secondary_skills : making.required_skills ) {
+                if( np->get_skill_level( secondary_skills.first ) < secondary_skills.second ) {
+                    has_all_skills = false;
+                }
+            }
+        }
+        if( has_all_skills && !making.character_has_required_proficiencies( *np ) ) {
+            has_all_skills = false;
+        }
+        if( has_all_skills ) {
+            const float helper_skill_average = std::max( np->get_recipe_weighted_skill_average( making ),
+                                               0.0f );
+            npc_helpers_weighted_average += std::pow( helper_skill_average, 2 );
+            add_msg_if_player( m_info, _( "%s helps with crafting…" ), np->name );
         }
     }
 
-    return ( skill_roll / diff_roll ) * prof_multiplier;
+    // Use a sum of squares to determine the final weighted average.
+    negative = false;
+    float summed_skills = player_weighted_skill_average + npc_helpers_weighted_average;
+    if( summed_skills < 0 ) {
+        negative = true;
+        summed_skills *= -1;
+    }
+    float weighted_skill_average = sqrt( summed_skills );
+    if( negative ) {
+        weighted_skill_average *= -1;
+    }
+
+    int secondary_difficulty = 0;
+    int secondary_level_count = 0;
+    for( const auto &count_secondaries : making.required_skills ) {
+        secondary_level_count += count_secondaries.second;
+        secondary_difficulty += std::pow( count_secondaries.second, 2 );
+    }
+    const float final_difficulty =
+        ( 2.0f * making.difficulty * making.difficulty + 1.0f * secondary_difficulty ) /
+        ( 2.0f * making.difficulty + 1.0f * secondary_level_count );
+
+    // in the future we might want to make the standard deviation vary depending on some feature of the recipe.
+    float crafting_stddev = 2.0f;
+    if( final_difficulty > weighted_skill_average ){
+        // Increase the standard deviation by 0.33 for every point below the difficulty your skill level is.
+        // This makes the rolls more random and "swingy" at low levels, based more on luck.
+        crafting_stddev += ( final_difficulty - weighted_skill_average ) / 3;
+    }
+    if( final_difficulty < weighted_skill_average ){
+        // decrease the standard deviation by 0.25 for every point above the difficulty your skill level is, to a cap of 1.
+        // This means that luck plays less of a role the more overqualified you are.
+        crafting_stddev -= std::min( ( weighted_skill_average - final_difficulty ) / 4, 1.0f );
+    }
+    float craft_roll = std::max( normal_roll( weighted_skill_average, crafting_stddev ), 0.0 );
+    
+    return std::max( craft_roll / final_difficulty, 0.0f );
```

Comment:
May I suggest we switch from a ratio of two normals to the difference of two normals? The ratio of two normals is well known to be a mess to work with. This would require some changes as the `max(A/B, 0)` would no longer be appropiate and success would have to be redefined as `success_roll > 0.0` instead of `success_roll > 1.0`, but it would have the advantage that we would have a behaving output distribution. This would enable relatively straightforward computation of probabilities so we could show to people in the crafting menu what their failure probability is.

Reply:
Yes, we should do that.

Reply:
I believe the reason it is presently done like this is to attempt to maintain the same range output results so I don't have to alter the failure case stuff. However it might be easy enough to do that with subtraction instead of division.

Reply:
Divide the result by the average output of the difficulty dice (or the skill dice if preferable), then add 1?

Reply:
There aren't really difficulty or skill dice anymore. there is a normal roll, centered around your average skill with bonuses/penalties, and a constant difficulty determined by all contributing skills.

It might be possible to just subtract A from B and apply it to current rolls. Presently they're based on a ratio of less than or greater than 1 though so I guess we probably need to add 1 to that final value such that if your success perfectly equals the difficulty, that's considered a pass.

Reply:
> There aren't really difficulty or skill dice anymore. there is a normal roll, centered around your average skill with bonuses/penalties, and a constant difficulty determined by all contributing skills.

Got it; was confused by how it was previously set up, sorry.

> It might be possible to just subtract A from B and apply it to current rolls. Presently they're based on a ratio of less than or greater than 1 though so I guess we probably need to add 1 to that final value such that if your success perfectly equals the difficulty, that's considered a pass.

How about (for the purpose of reusing existing code that doesn't seem to be problematic, like getting proportions of lost components and progress, and potentially for figuring out any additional advantages of a very successful roll, like things getting done faster than normal):
- Take roll - difficulty:
- - If this is above 0, divide by the difficulty. (Thus, the greater the difficulty, the harder it is to get any additional advantages.)
- - If this is below 0, divide by the _mean_ result of the roll. (Thus, the better your skill, the more limited any disaster due to mischance is.)
- Add 1.
- Take max(this, 0).

Reply:
The code actually already accounts for your likelihood of avoiding disaster at higher levels by reducing the standard deviation, and expanding it at skill deficit. so at high levels you get a progressively less broad spread of results.

I'll have to review the rest of your math there. I think I'd prefer to have a simple calculation and go through and edit how it's used elsewhere than have a complex calculation that can be plugged I to the existing code seemlessly though.